File: cpp/task.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#define FILE_REMAINING "task.txt"
#define FILE_COMPLETED "completed.txt"
using std::cin;
using std::cout;
using std::ifstream;
using std::ofstream;
using std::string;
using std::vector;

void help()
{
    cout << "Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics";
}

vector<string> getFileData(string fileName)
{
    int ndx = -1;
    ifstream fin;
    fin.open(fileName);
    vector<string> data;
    string line;
    while (fin)
    {
        getline(fin, line);
        if (line.empty())
            continue;
        ndx++;
        data.insert(data.begin() + ndx, line);
    }
    fin.close();
    return data;
}

void setFileData(vector<string> data, string fileName)
{
    ofstream fout;
    fout.open(fileName);
    for (int i = 0; i < data.size(); i++)
        fout << data[i] << std::endl;

    fout.close();
}

void getRemainingTasks(string fileName, bool displayQuantity)
{
    vector<string> data = getFileData(fileName);
    if (data.empty() && !displayQuantity)
    {
        cout << "There are no pending tasks!";
        return;
    }
    if (displayQuantity)
        cout << (fileName == FILE_REMAINING ? "Pending : " : "Completed : ") << data.size() << std::endl;

    for (int i = 0; i < data.size(); i++)
    {
        string item = data[i];
        cout << (fileName == FILE_REMAINING ? std::to_string((i + 1)) + ". " + item.substr(item.find(" ") + 1) + " [" + item.substr(0, item.find(" ")) + "]" : std::to_string((i + 1)) + ". " + item) << std::endl;
    }
}

void setNewTask(string priority, string task)
{
    vector<string> data = getFileData(FILE_REMAINING);
    int priorityInt;
    try
    {
        priorityInt = std::stoi(priority);
    }
    catch (...)
    {
        cout << "Invalid priority";
        return;
    }
    int ndx = 0;
    for (int i = 0; i < data.size(); i++)
    {
        int cur;
        try
        {
            cur = std::stoi(data[i].substr(0, data[i].find(" ")));
        }
        catch (...)
        {
        }
        if (priorityInt < cur)
            break;
        ndx++;
    }
    data.insert(data.begin() + ndx, priority + " " + task);
    setFileData(data, FILE_REMAINING);
    cout << "Added task: \"" << task << "\" with priority " << priority;
}

string deleteTask(string index)
{
    vector<string> data = getFileData(FILE_REMAINING);
    int ndx;
    try
    {
        ndx = std::stoi(index);
    }
    catch (...)
    {
        cout << "Enter a valid index";
        return "";
    }
    if (ndx < 1 || ndx > data.size())
        return "";
    string deletedItem = data[ndx - 1];
    deletedItem = deletedItem.substr(deletedItem.find(" ") + 1);
    data.erase(data.begin() + ndx - 1);
    setFileData(data, FILE_REMAINING);
    return deletedItem;
}

void setCompleted(string index)
{
    string deletedItem = deleteTask(index);
    if (deletedItem.empty())
    {
        cout << "Error: no incomplete item with index #" << index << " exists.";
        return;
    }

    ofstream fout;
    fout.open(FILE_COMPLETED, std::ios::app);
    fout << deletedItem << std::endl;
    fout.close();
    cout << "Marked item as done.";
}

void removeExtraSpaces(string *str)
{
    string st = *str;
    string s = "  ", t = " ";
    string::size_type n = 0;
    while ((n = st.find(s)) != string::npos)
        st.replace(n, s.size(), t);
    if (st.size() > 0)
    {
        if (st.substr(0, 1) == " ")
            st.replace(0, 1, "");
        if (st.size() > 0)
        {
            if (st.substr(st.size() - 1) == " ")
                st.replace(st.size() - 1, 1, "");
        }
    }
    *str = st;
}

int main(int argc, char *argv[])
{
    string primaryArg = argc > 1 ? argv[1] : "";
    if (primaryArg == "ls")
        getRemainingTasks(FILE_REMAINING, false);
    else if (primaryArg == "add")
    {
        if (argc > 3)
        {
            string arg = argv[2];
            string task = argv[3];
            if (task.find(" ") != string::npos)
                removeExtraSpaces(&task);
            if (task.empty())
                cout << "Error: Missing tasks string. Nothing added!";
            else
            {
                if (arg.compare("0") >= 0)
                    setNewTask(arg, task);
                else
                    cout << "Invalid priority";
            }
        }
        else
            cout << "Error: Missing tasks string. Nothing added!";
    }
    else if (primaryArg == "del")
    {
        if (argc > 2)
        {
            if (deleteTask(argv[2]).empty())
                cout << "Error: task with index #" << argv[2] << " does not exist. Nothing deleted.";
            else
                cout << "Deleted task #" << argv[2];
        }
        else
            cout << "Error: Missing NUMBER for deleting tasks.";
    }
    else if (primaryArg == "done")
    {
        if (argc > 2)
            setCompleted(argv[2]);
        else
            cout << "Error: Missing NUMBER for marking tasks as done.";
    }
    else if (primaryArg == "report")
    {
        getRemainingTasks(FILE_REMAINING, true);
        cout << std::endl;
        getRemainingTasks(FILE_COMPLETED, true);
    }
    else
        help();

    return 0;
}

File: javascript/task.js
const args = process.argv
const fs = require('fs')
const FILE_REMAINING = 'task.txt'
const FILE_COMPLETED = 'completed.txt'

fs.openSync(FILE_REMAINING, 'a+')
fs.openSync(FILE_COMPLETED, 'a+')

const help = () => {
    process.stdout.write("Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics")
}

const getRemainingTasks = (fileName, displayQuantity) => {
    let data = fs.readFileSync(fileName, 'utf-8')
    data = data.split("\n")
    if (data[data.length - 1] == '')
        data.splice(data.length - 1, 1)

    if (displayQuantity) {
        if (fileName == FILE_REMAINING)
            console.log(`Pending : ${data.length}`)
        else
            console.log(`Completed : ${data.length}`)
    }else if (data.length == 0) {
        console.log("There are no pending tasks!")
        return
    }

    let item
    for (let i = 0; i < data.length; i++) {
        item = data[i]
        if (item) {
            if (fileName == FILE_REMAINING)
                console.log(`${i + 1}. ${item.substring(item.indexOf(' ') + 1)} [${item.substring(0, item.indexOf(' '))}]`)
            else
                console.log(`${i + 1}. ${item}`)
        }
    }
}

const setNewTask = (priority, task) => {
    let data = fs.readFileSync(FILE_REMAINING, 'utf-8')
    data = data.split("\n")
    let ndx = 0
    data.forEach(item => {
        if (priority < item.substring(0, item.indexOf(' ')))
            return
        ndx++
    })

    data.splice(ndx, 0, `${priority} ${task}`)

    let newData = ''
    for (let i = 0; i < data.length; i++) {
        if (data[i] == '')
            continue
        newData += i < data.length - 1 ? `${data[i]}\n` : data[i]
    }

    fs.writeFileSync(FILE_REMAINING, newData)
    console.log(`Added task: "${task}" with priority ${priority}`);
}
const deleteTask = (index) => {
    let data = fs.readFileSync(FILE_REMAINING, 'utf-8')
    data = data.split("\n")
    if (index < 1 || index > data.length || data[0] == '') {
        return
    }
    let deletedItem = data[index - 1]
    deletedItem = deletedItem.substring(deletedItem.indexOf(' ') + 1)
    data.splice(index - 1, 1)

    let newData = ''
    for (let i = 0; i < data.length; i++) {
        if (data[i] == '')
            continue
        newData += i < data.length - 1 ? `${data[i]}\n` : data[i]
    }

    fs.writeFileSync(FILE_REMAINING, newData)

    return deletedItem
}

const setCompleted = (index) => {
    const deletedItem = deleteTask(index)
    if (!deletedItem) {
        console.log(`Error: no incomplete item with index #${index} exists.`);
        return
    }
    fs.appendFileSync(FILE_COMPLETED, `${deletedItem}\n`)
    console.log("Marked item as done.")
}

switch (args[2]) {
    case 'help':
        help()
        break;
    case 'ls':
        getRemainingTasks(FILE_REMAINING);
        break;
    case 'add':
        if (args[3] && args[4]) {
            if (args[3] >= '0')
                setNewTask(args[3], args[4])
            else
                console.log('Priority cannot be negative')
        }
        else
            console.log("Error: Missing tasks string. Nothing added!")
        break;
    case 'del':
        if (args[3]) {
            if (deleteTask(args[3]))
                console.log(`Deleted task #${args[3]}`);
            else
                console.log(`Error: task with index #${args[3]} does not exist. Nothing deleted.`);
        }
        else
            console.log("Error: Missing NUMBER for deleting tasks.")
        break;
    case 'done':
        if (args[3])
            setCompleted(args[3])
        else
            console.log("Error: Missing NUMBER for marking tasks as done.")
        break;
    case 'report':
        getRemainingTasks(FILE_REMAINING, true)
        console.log()
        getRemainingTasks(FILE_COMPLETED, true)
        break;
    default:
        help()
}